pkg/chunked/storage_linux.go

```diff
@@ -1734,3 +1754,63 @@ func validateChunkChecksum(chunk *internal.FileMetadata, root, path string, offs
 
 	return digester.Digest() == digest
 }
+
+func newStagedFileGetter(dirFD int, differOpts *graphdriver.DifferOptions, entriesWithOriginalNames []fileMetadata) (*stagedFileGetter, error) {
+	var pathMap map[string]string = nil
+	if differOpts != nil && differOpts.Format == graphdriver.DifferOutputFormatFlat {
+		pathMap = make(map[string]string, len(entriesWithOriginalNames))
+		// This needs to match cloneEntriesWithFlatNames
+		for _, e := range entriesWithOriginalNames {
+			if e.Type != TypeReg {
+				continue
+			}
+			if e.Digest == "" {
+				return nil, fmt.Errorf("missing digest for %q", e.Name)
+			}
+			digest, err := digest.Parse(e.Digest)
+			if err != nil {
+				return nil, err
+			}
+			d := digest.Encoded()
+			pathMap[filepath.Clean(e.Name)] = fmt.Sprintf("%s/%s", d[0:2], d[2:])
+		}
+	}
+
+	dirFD2, err := unix.Dup(dirFD)
+	if err != nil {
+		return nil, err
+	}
+	rootDir := os.NewFile(uintptr(dirFD2), "staging root directory")
+	if rootDir == nil {
+		return nil, fmt.Errorf("failed creating a directory os.File")
+	}
+	return &stagedFileGetter{
+		rootDir: rootDir,
+		pathMap: pathMap,
+	}, nil
+}
+
+type stagedFileGetter struct {
+	rootDir *os.File
+	pathMap map[string]string // nil, or a map from filepath.Clean()ed tar file names to expected on-filesystem names
+}
+
+func (fg *stagedFileGetter) Get(filename string) (io.ReadCloser, error) {
+	if fg.pathMap != nil {
+		path, ok := fg.pathMap[filepath.Clean(filename)]
+		if !ok {
+			return nil, fmt.Errorf("no path mapping exists for tar entry %q", filename)
+		}
+		filename = path
+	}
+	pathFD, err := securejoin.OpenatInRoot(fg.rootDir, filename)
```

Comment:
(A little surprising securejoin doesn't have a high level `securejoin.OpenReader` or something)

Reply:
(We have an `userns.secureOpen` internally with two other callers — centralizing that, and proposing that upstream, does sound attractive.)